### ISSUE TITLE
Real Time Global/Indirect Illumination

### ISSUE THREAD

**Issue by AlexeyGapon** [mrdoob/three.js#6575]
Are you planning to implement Real Time Global/Indirect illumination shadows for Three.js  renderer?

**Comment by WestLangley** [comment]
For static scenes, you can bake the indirect illumination into a light map. As light maps are now purely additive to total light in r.72dev, this should work as expected.

For dynamic scenes, a different technique would have to be implemented. We are open to suggestions.

/ping @bhouston

**Comment by bhouston** [comment]
True global illumination in real-time is super hard.

But if you wanted to bake lights, we handle it as described by @WestLangley.  I'd suggest going that way for the time being.

**Comment by AlexeyGapon** [comment]
Im want to write a renderer for my project, now i am currently workin on it. 
It will have DOF, GODRAYS, GI.  So i am working on inbox composer for it.
it will be like
var  scene = new THREE.Scene(); 
var camera = ...blah;
var renderer = new THREE.WebGLAdvancedRenderer();
renderer.Add.GodRaysEmitter ([array of lights]);
renderer.Add.dof({patams});


renderer.render(scene, camera);
I am sking about GO because i know there are real time GI rendering tekniques.

**Comment by bhouston** [comment]
DOF is really easy and God Rays are moderately easy -- there are ThreeJS examples for those things.

Real-time GI is more complex - there are a few options and none of them are that easy to implement.  
Until Unity 5, it relied upon light baking, and I'd suggest that route.

It Unity 5 uses Enlighten by Geomerics.  I've asked them and it is not yet available for JavaScript, but Enlighten is pretty frigging awesome (and unfortunately likely covered by patents as well):

https://www.youtube.com/watch?v=VHbHOQ1NRuw&feature=youtu.be

**Comment by bhouston** [comment]
Here is Enlighten in Unity 5 --- it is pretty awesome:

https://www.youtube.com/watch?v=ouJNRJ2uPmY

Here is a SIGGRAPH 2010 talk that describes how Geometrics does real-time radiosity:

https://www.youtube.com/watch?v=h2lbZC3ihUU

**Comment by RemusMar** [comment]
> True global illumination in real-time is super hard.

For a complex high-poly scene is not an option (it's a CPU/GPU killer).
Anyway, forget about "true" GI.
For static lights & objects, you can get similar results with "smart" Light/Shadow maps.
Not to mention that you'll get much better performances.

cheers

**Comment by AlexeyGapon** [comment]
i think guys from nvidia shoud create hardvare GI aceleration)

**Comment by mattdesl** [comment]
This is a super cool area that I'm keen to explore. 

Here's one of the nicer implementations I've seen:
http://graphics.cs.williams.edu/papers/PhotonHPG09/

The algorithm is implemented with async PBO texture upload (not supported in WebGL), CPU raycasting spread across an 8-core processor, and some other desktop-centric things. So I'm really not sure it's at all feasible in WebGL. 

If the plan is for ThreeJS to be the forerunner of high-end WebGL rendering, then I think it would be best to just provide the developer with a G-buffer render pipeline that extensions can build on. There's lots of other cool things that could be built on this, like [Aggregate G-Buffer Anti-Aliasing](http://graphics.cs.williams.edu/papers/AggregateI3D15/).

**Comment by AlexeyGapon** [comment]
Look what i found. All source is there! 
http://madebyevan.com/webgl-path-tracing/

**Comment by AlexeyGapon** [comment]
I it possible!
Just do not calculate IrMap each frame! 
1) Separate models to static and dinamic
2)  Calculate GI 4 times smaller then resolution then just blur it.
3) Do not calculate all world but only some distance form dinamic obejcts that currently moving.

**Comment by Usnul** [comment]
#5942
#4434
@AlexeyGapon that path tracing by evan is great, except it's not usable directly. Path tracing produces photo-realistic results by mimicking photon behaviour, however - it's not practical yet. Not even brigade can claim to be usable for games on your average PC yet. 

Distance Field Ambient Occlusion is an interesting topic, something that's fairly easy to grasp. It's not GI per se, but it does give shadowing. 
https://docs.unrealengine.com/latest/INT/Engine/Rendering/LightingAndShadows/DistanceFieldAmbientOcclusion/index.html

here's a 3 year-old implementation of GI by @pyalot for WebGL
http://codeflow.org/entries/2012/aug/25/webgl-deferred-irradiance-volumes/
looks quite breath-taking. Even if not quite "automated".

**Comment by AlexeyGapon** [comment]
I sow the last example. And i think we defenetly need something like that in our rendering engines. 
Now i am trying to implement godrays with SSAO and DoF it is already look nice. But if we could use something even more powerfull. I am ready to be a part of devlopment team if anybody is interested. Course I am a lot!

**Comment by AlexeyGapon** [comment]
i am suggesting not the real gi but something that almost similar with looking on it. I mean we could try to claculate voxeled "GI" world  for example.  We have static meshes and dinamic meshes. Static meshes produces it`s reflexes only once befor first render frame is shown. Then we will calculate the VoxelGI world for dinamic objects. And once there transform changed we just update the voxelGI world with reflexes of the dynamic object reflexes of the dynamic obects are consts so we can calculate them only once too. So when the obect is mooving we just recalculate the VoxelGI world only for some distance from the moving object.  So we will have the matrix of the world, matrix for each dynamic object and once this object is moving we will recalculate the displaying matrix of reflexes.

**Comment by mrdoob** [comment]
This is an exaggeration but, evan's gi pool only works for pools that look exactly like his pool and only if evan is coding/creating the assets.

**Comment by pyalot** [comment]
I'd like to note that the deferred irradiance technique is quite limited, and in every case that I was asked to use it in a project so far, it was ultimately not used because it was too authoring intensive.

**Comment by bhouston** [comment]
More developments in this area:

http://www.aduprat.com/portfolio/?page=articles%2FPBGI

http://oxidegames.com/2016/03/19/object-space-lighting-following-film-rendering-2-decades-later-in-real-time/
